File: cpp/src/tensor.cpp
#include "../include/tensor.h"
#include <vector>
#include <string>
#include <numeric>
#include <algorithm>
#include <cstring>

namespace minitorch {

struct Tensor::Impl {
    std::vector<int64_t> shape;
    std::string dtype;
    std::string device;
    std::vector<float> data;
};

Tensor::Tensor(const std::vector<int64_t>& shape, const std::string& dtype, const std::string& device) {
    impl_ = std::make_shared<Impl>();
    impl_->shape = shape;
    impl_->dtype = dtype;
    impl_->device = device;
    int64_t size = 1;
    for (auto d : shape) size *= d;
    impl_->data.resize(size, 0.0f);
}

Tensor::~Tensor() {}

void* Tensor::data() { return impl_->data.data(); }
std::vector<int64_t> Tensor::shape() const { return impl_->shape; }
std::string Tensor::dtype() const { return impl_->dtype; }
std::string Tensor::device() const { return impl_->device; }

Tensor Tensor::add(const Tensor& other) const {
    Tensor out(impl_->shape, impl_->dtype, impl_->device);
    for (size_t i = 0; i < impl_->data.size(); ++i) {
        out.impl_->data[i] = impl_->data[i] + other.impl_->data[i];
    }
    return out;
}

Tensor Tensor::mul(const Tensor& other) const {
    Tensor out(impl_->shape, impl_->dtype, impl_->device);
    for (size_t i = 0; i < impl_->data.size(); ++i) {
        out.impl_->data[i] = impl_->data[i] * other.impl_->data[i];
    }
    return out;
}

Tensor Tensor::matmul(const Tensor& other) const {
    // Only supports 2D for demo
    int64_t m = impl_->shape[0];
    int64_t k = impl_->shape[1];
    int64_t n = other.impl_->shape[1];
    Tensor out({m, n}, impl_->dtype, impl_->device);
    for (int64_t i = 0; i < m; ++i) {
        for (int64_t j = 0; j < n; ++j) {
            float sum = 0.0f;
            for (int64_t l = 0; l < k; ++l) {
                sum += impl_->data[i * k + l] * other.impl_->data[l * n + j];
            }
            out.impl_->data[i * n + j] = sum;
        }
    }
    return out;
}

Tensor Tensor::sum() const {
    float s = std::accumulate(impl_->data.begin(), impl_->data.end(), 0.0f);
    Tensor out({1}, impl_->dtype, impl_->device);
    out.impl_->data[0] = s;
    return out;
}

} // namespace minitorch

File: python/minitorch/tensor.py
import numpy as np
from typing import Any, Tuple, Optional

class Tensor:
    """
    Tensor class wrapping C++ core. Supports creation, elementwise ops, shape, device, and autograd hooks.
    """
    def __init__(self, data, requires_grad=False, device="cpu", dtype=None, _cpp_tensor=None):
        if _cpp_tensor is not None:
            self._data = _cpp_tensor
            self.shape = tuple(self._data.shape())
            self.device = self._data.device()
            self.dtype = self._data.dtype()
        else:
            self._data = np.array(data, dtype=dtype)
            self.shape = self._data.shape
            self.device = device
            self.dtype = self._data.dtype
        self.requires_grad = requires_grad
        self.grad = None
        self._grad_fn = None
        self._ctx = None

    @staticmethod
    def ones(*shape, requires_grad=False, device="cpu", dtype=None):
        data = np.ones(shape, dtype=dtype)
        return Tensor(data, requires_grad, device, dtype)

    @staticmethod
    def zeros(*shape, requires_grad=False, device="cpu", dtype=None):
        data = np.zeros(shape, dtype=dtype)
        return Tensor(data, requires_grad, device, dtype)

    @staticmethod
    def randn(*shape, requires_grad=False, device="cpu", dtype=None):
        data = np.random.randn(*shape).astype(dtype or np.float32)
        return Tensor(data, requires_grad, device, dtype or np.float32)

    def numpy(self):
        return self._data

    def __add__(self, other):
        if isinstance(other, Tensor):
            out = Tensor(self._data + other._data)
        else:
            out = Tensor(self._data + other)
        return out

    def __mul__(self, other):
        if isinstance(other, Tensor):
            out = Tensor(self._data * other._data)
        else:
            out = Tensor(self._data * other)
        return out

    def __matmul__(self, other):
        if isinstance(other, Tensor):
            out = Tensor(self._data @ other._data)
        else:
            out = Tensor(self._data @ other)
        return out

    def sum(self):
        return Tensor(self._data.sum())

    def backward(self, grad=None):
        from .autograd import backward
        backward(self, grad)

    def to(self, device):
        # For now, only CPU supported
        return self

    def __repr__(self):
        return f"Tensor({self._data}, requires_grad={self.requires_grad}, device={self.device})"
